lib/ansible/plugins/action/gather_facts.py

```diff
@@ -0,0 +1,110 @@
+
+            fact_filter = mod_args.pop('filter', None)
+            if fact_filter is not None:
+                self._display.warning('Ignoring filter(%s) for %s' % (fact_filter, fact_module))
+
+        return mod_args
+
+    def run(self, tmp=None, task_vars=None):
+
+        self._supports_check_mode = True
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        result['ansible_facts'] = {}
+
+        modules = C.config.get_config_value('FACTS_MODULES', variables=task_vars)
+        parallel = task_vars.pop('ansible_facts_parallel', self._task.args.pop('parallel', None))
+
+        if 'smart' in modules:
+            connection_map = C.config.get_config_value('CONNECTION_FACTS_MODULES', variables=task_vars)
+            modules.extend([connection_map.get(self._connection._load_name, 'setup')])
+            modules.pop(modules.index('smart'))
+
+        failed = {}
+        skipped = {}
+        if parallel is False or (len(modules) == 1 and parallel is None):
+            # serially execute each module
+            for fact_module in modules:
+                # just one module, no need for fancy async
+                mod_args = self._get_module_args(fact_module, task_vars)
+                res = self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=False)
+                if res.get('failed', False):
+                    failed[fact_module] = res.get('msg')
+                elif res.get('skipped', False):
+                    skipped[fact_module] = res.get('msg')
+                else:
+                    result = combine_vars(result, {'ansible_facts': res.get('ansible_facts', {})})
+        else:
+            # do it async
+            jobs = {}
+            for fact_module in modules:
+
+                mod_args = self._get_module_args(fact_module, task_vars)
+                self._display.vvvv("Running %s" % fact_module)
+                jobs[fact_module] = (self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=True))
+
+            while jobs:
+                for module in jobs:
+                    poll_args = {'jid': jobs[module]['ansible_job_id'], '_async_dir': os.path.dirname(jobs[module]['results_file'])}
+                    res = self._execute_module(module_name='async_status', module_args=poll_args, task_vars=task_vars, wrap_async=False)
+                    if 'finished' in res:
+                        if res.get('failed', False):
+                            failed[module] = res.get('msg')
+                        elif res.get('skipped', False):
+                            skipped[module] = res.get('msg')
+                        else:
+                            result = combine_vars(result, {'ansible_facts': res.get('ansible_facts', {})})
+                        del jobs[module]
+                        break
+                else:
+                    time.sleep(0.5)
```

Comment:
I believe this `else` isn't indented far enough.  I think it needs to be indented 1 further, as it is, there is no wait between `async_status` polling, causing very high load on the controller.

Reply:
was meant to keep a delay between the batches of job polling, im not sure we want to throttle the polling of multiple plugins the same way, going to add one but at a smaller sleep